numpy/lib/format.py

```diff
@@ -623,13 +623,22 @@ def _read_array_header(fp, version, max_header_size=_MAX_HEADER_SIZE):
     #   "descr" : dtype.descr
     # Versions (2, 0) and (1, 0) could have been created by a Python 2
     # implementation before header filtering was implemented.
-    if version <= (2, 0):
-        header = _filter_header(header)
+    #
+    # For performance reasons, we try without _filter_header first though
     try:
         d = safe_eval(header)
     except SyntaxError as e:
-        msg = "Cannot parse header: {!r}"
-        raise ValueError(msg.format(header)) from e
+        try:
+            if version <= (2, 0):
+                d = safe_eval(_filter_header(header))
+                warnings.warn("Read array created with Python 2, requires " +
+                "header filtering. Consider saving with Python 3 for " +
+                "faster loading.", UserWarning, stacklevel=2)
+            else:
+                raise e
+        except SyntaxError as e:
+            msg = "Cannot parse header: {!r}"
+            raise ValueError(msg.format(header)) from e
```

Comment:
I think this has better exception chaining semantics:
```suggestion
        if version <= (2, 0):
            header = _filter_header(header)
            try:
                d = safe_eval(header)
            except SyntaxError as e2:
                msg = "Cannot parse header: {!r}"
                raise ValueError(msg.format(header)) from e2
            else:
                warnings.warn("Read array created with Python 2, requires " +
                "header filtering. Consider saving with Python 3 for " +
                "faster loading.", UserWarning, stacklevel=2)
        else:
            msg = "Cannot parse header: {!r}"
            raise ValueError(msg.format(header)) from e
```

Reply:
Wow, love those fast comments, thank you a lot!

I know, my PR might look a little bit weird, but using the version you proposed will almost always trigger `_filter_header`, as Numpy arrays are saved as version 1 probably most of the time (unless the header is very large or contains Unicode). However, `_filter_header` is a bottleneck, compare https://github.com/numpy/numpy/issues/22898

However, `_filter_header` is only necessary if the file was created by Python 2, as strange things happen there with numbers.

Therefore, triggering `_filter_header` should be a fallback rather than a default.

Reply:
Sorry, I think I misunderstood that part. It can be done that way, indeed. Let me clean it up.